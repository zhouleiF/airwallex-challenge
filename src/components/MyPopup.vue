<template>
  <div class="popup" @touchmove.prevent @click.self="close" v-show="visible">
    <div class="con">
      <div class="title">
        <slot name="title" />
        <div class="line"></div>
      </div>
      <div class="content">
        <slot name="content" />
      </div>
      <div class="footer">
        <slot name="footer" />
      </div>
    </div>
  </div>
</template>
<script>
export default {
  name: 'MyPopup',
  data () {
    return {
      visible: false
    }
  },
  methods: {
    close () {
      this.$emit('beforeClose')
      this.visible = false
    },
    show () {
      this.visible = true
    }
  }
}
</script>
<style lang="scss" scoped>
  .popup {
    box-sizing: content-box;
    position: fixed;
    display: flex;
    justify-content: center;
    align-items: center;
    z-index: 9999;
    top: 0;
    bottom: 0;
    left: 0;
    right: 0;
    background-color: rgba(0,0,0,.5);
    .con {
      width: 600px;
      min-height: 300px;
      background-color: #fff;
      border-radius: 10px;
      padding: 70px 0 40px;
      .title {
        font-size: 35px;
        color: #5C6C78;
        line-height: 1.2;
      }
      .line {
        width: 100px;
        height: 1PX;
        background-color: #000;
        margin: 20px auto 0;
      }
      .content {
        padding-bottom: 30px;
        display: flex;
        justify-content: center;
        align-items: center;
        text-align: center;
      }
      .footer {
        text-align: center;
        button {
          display: block;
          width: 550px;
          height: 90px;
          line-height: 90px;
          font-size: 42px;
          text-align: center;
          color: #56BACF;
          margin: 0 auto 50px;
          background-color: transparent;
          &:first-child {
            color: #fff;
            border-radius: 50px;
            background-color: #56BACF;
          }
          &:last-child {
            margin-bottom: 0
          }
        }
      }
    }
  }
</style>
