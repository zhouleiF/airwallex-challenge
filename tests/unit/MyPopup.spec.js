import { mount } from '@vue/test-utils'
import MyPopup from '@/components/MyPopup.vue'

const title = 'this is popup title'
const content = 'this is popup content'
const footer = 'this is popup footer'
const beforeClose = jest.fn(() => {})

describe('MyPopup component', () => {
  const wrapper = mount(MyPopup, {
    slots: {
      title,
      content,
      footer
    },
    listeners: {
      beforeClose
    }
  })
  it('renders correctly', () => {
    expect(wrapper.element).toMatchSnapshot()
  })
  it('renders title slot', () => {
    expect(wrapper.find('.title').text()).toEqual(title)
  })
  it('renders content slot', () => {
    expect(wrapper.find('.content').text()).toEqual(content)
  })
  it('renders footer slot', () => {
    expect(wrapper.find('.footer').text()).toEqual(footer)
  })
  it('default is hidden', () => {
    expect(wrapper.isVisible()).toBe(false)
  })
  it('show the popup', async () => {
    wrapper.vm.show()
    await wrapper.vm.$nextTick()
    expect(wrapper.isVisible()).toBe(true)
  })
  it('hide the popup by calling close method', async () => {
    wrapper.vm.close()
    await wrapper.vm.$nextTick()
    expect(wrapper.isVisible()).toBe(false)
  })
  it('hide the popup by click the background ', async () => {
    wrapper.vm.show()
    await wrapper.vm.$nextTick()
    expect(wrapper.isVisible()).toBe(true)
    wrapper.trigger('click')
    await wrapper.vm.$nextTick()
    expect(wrapper.isVisible()).toBe(false)
  })
  it('emit beforeClose event when popup hidden, we have tried two times', () => {
    expect(beforeClose.mock.calls.length).toBe(2)
  })
})
