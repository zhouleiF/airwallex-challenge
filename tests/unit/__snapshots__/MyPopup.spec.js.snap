// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`MyPopup component renders correctly 1`] = `
<div
  class="popup"
  style="display: none;"
>
  <div
    class="con"
  >
    <div
      class="title"
    >
      this is popup title 
      <div
        class="line"
      />
    </div>
     
    <div
      class="content"
    >
      this is popup content
    </div>
     
    <div
      class="footer"
    >
      this is popup footer
    </div>
  </div>
</div>
`;
